File: solver.js
/*
 * Matrix Position Solver -- core algorithm used by the GitHub Pages app
 * (index.html). Kept in its own file so it can be loaded by the page AND
 * mirrored by solver.py for unit testing (test_solver.py). The Python twin is
 * the line-for-line equivalent; keep the two in sync.
 *
 * Puzzle: each active row holds one element at position 1..7. A move picks a
 * row and direction; the moved row shifts and dependent rows shift per the
 * dependency matrix:
 *   deps[r][j] === 'x'  -> moving row r shifts row j the SAME direction
 *   deps[r][j] === 'o'  -> ... the OPPOSITE direction
 *   deps[r][j] === '.'  -> row j unaffected by moving row r
 * direction 'left' = +1 (toward 7), 'right' = -1 (toward 1).
 * A move is blocked if it would push ANY element outside [1, 7].
 * Goal: bring every active element to position 4 in the fewest moves.
 */
(function (root, factory) {
  var api = factory();
  if (typeof module !== 'undefined' && module.exports) module.exports = api;
  else root.MatrixSolver = api;
})(typeof globalThis !== 'undefined' ? globalThis : this, function () {
  'use strict';

  var MIN_POS = 1, MAX_POS = 7, TARGET = 4;

  // Self-contained move (reads deps directly). This is the public API the UI
  // uses to render the state after each step. solve() uses the precomputed
  // fast path below; test_solver checks the two agree.
  function applyMove(state, row, dir, deps, n) {
    var delta = dir === 'left' ? 1 : -1;
    var ns = state.slice();

    if (state[row] !== 0) {
      var p = state[row] + delta;
      if (p < MIN_POS || p > MAX_POS) return null;
      ns[row] = p;
    }

    for (var j = 0; j < n; j++) {
      if (j === row || state[j] === 0) continue;
      var dep = deps[row][j];
      if (dep === 'x') {
        var px = state[j] + delta;
        if (px < MIN_POS || px > MAX_POS) return null;
        ns[j] = px;
      } else if (dep === 'o') {
        var po = state[j] - delta;
        if (po < MIN_POS || po > MAX_POS) return null;
        ns[j] = po;
      }
    }

    return ns;
  }

  // Per row, the list of [index, sign] it shifts when moved: sign +1 for "same"
  // (the row itself and 'x' deps), -1 for 'o'. Computed once so each BFS
  // expansion walks only affected indices instead of rescanning deps strings.
  function buildEffects(deps, n) {
    var effects = [];
    for (var r = 0; r < n; r++) {
      var rowEff = [[r, 1]];
      for (var j = 0; j < n; j++) {
        if (j === r) continue;
        if (deps[r][j] === 'x') rowEff.push([j, 1]);
        else if (deps[r][j] === 'o') rowEff.push([j, -1]);
      }
      effects.push(rowEff);
    }
    return effects;
  }

  // Pack a state (each cell 0..7) into one integer so it can key a Map/Set far
  // faster than a joined string. Up to 7 rows -> max 8^7, a safe integer.
  function encode(state, n) {
    var k = 0;
    for (var i = 0; i < n; i++) k = k * 8 + state[i];
    return k;
  }

  function reconstruct(parent, endKey) {
    var moves = [];
    var k = endKey;
    var info = parent.get(k);
    while (info) {
      moves.push([info.row, info.dir]);
      k = info.prevKey;
      info = parent.get(k);
    }
    moves.reverse();
    return moves;
  }

  // Breadth-first search => shortest move list. Memory is O(states): one parent
  // pointer per discovered state, path rebuilt once at the end (the old code
  // copied a growing move array into every queued node).
  function solve(state, deps, n, active) {
    var start = state.slice();
    if (active.every(function (i) { return start[i] === TARGET; })) return [];

    var effects = buildEffects(deps, n);
    var dirs = [['left', 1], ['right', -1]];

    var parent = new Map();            // key -> {prevKey, row, dir}; start -> null
    parent.set(encode(start, n), null);
    var queue = [start];
    var head = 0;
    var MAX_STATES = 5000000;

    while (head < queue.length) {
      var cur = queue[head++];
      var curKey = encode(cur, n);

      for (var r = 0; r < n; r++) {
        var rowEff = effects[r];
        for (var di = 0; di < dirs.length; di++) {
          var dir = dirs[di][0], delta = dirs[di][1];

          var ns = cur.slice();
          var blocked = false;
          for (var e = 0; e < rowEff.length; e++) {
            var idx = rowEff[e][0];
            if (cur[idx] === 0) continue;
            var p = cur[idx] + rowEff[e][1] * delta;
            if (p < MIN_POS || p > MAX_POS) { blocked = true; break; }
            ns[idx] = p;
          }
          if (blocked) continue;

          var k = encode(ns, n);
          if (parent.has(k)) continue;
          parent.set(k, { prevKey: curKey, row: r, dir: dir });

          var solved = true;
          for (var a = 0; a < active.length; a++) {
            if (ns[active[a]] !== TARGET) { solved = false; break; }
          }
          if (solved) return reconstruct(parent, k);

          queue.push(ns);
          if (parent.size > MAX_STATES) return null;
        }
      }
    }
    return null;
  }

  return {
    applyMove: applyMove,
    solve: solve,
    encode: encode,
    buildEffects: buildEffects,
    MIN_POS: MIN_POS,
    MAX_POS: MAX_POS,
    TARGET: TARGET
  };
});

File: solver.py
"""Reference implementation of the Matrix Position Solver.

This is the Python twin of ``solver.js`` (the module the GitHub Pages app
actually loads). The deployed app runs the JavaScript version; this file exists
so the algorithm can be unit-tested with ``python test_solver.py`` -- there is
no JS runtime in this environment, so Python is the test harness. The two
implementations are intentionally line-for-line equivalent. If you change one,
change the other, and ``test_solver.py`` has a parity test to catch drift.

The puzzle
----------
Each active row holds one element at a position in ``1..7``. A move picks a row
and a direction. The chosen row shifts, and dependent rows shift according to
the dependency matrix:

    deps[r][j] == 'x'  -> moving row r shifts row j the SAME direction
    deps[r][j] == 'o'  -> moving row r shifts row j the OPPOSITE direction
    deps[r][j] == '.'  -> row j is unaffected by moving row r

Directions: ``'left'`` = +1 (toward position 7), ``'right'`` = -1 (toward 1).

A move is *blocked* (illegal) if it would push ANY element outside ``[1, 7]``.
Goal: bring every active element to position 4 in the fewest moves. ``solve``
returns the shortest move list (it is a breadth-first search), ``[]`` if the
puzzle is already solved, or ``None`` if no solution exists.
"""
from collections import deque

MIN_POS, MAX_POS, TARGET = 1, 7, 4


def apply_move(state, row, direction, deps, n):
    """Apply one move and return the new state tuple, or ``None`` if blocked.

    Self-contained (reads ``deps`` directly) so it mirrors the JS public API and
    is what the UI uses to render intermediate states. ``solve`` uses the
    precomputed-effects fast path below; ``test_solver`` checks they agree.
    """
    delta = 1 if direction == "left" else -1
    new_state = list(state)

    if state[row] != 0:
        p = state[row] + delta
        if p < MIN_POS or p > MAX_POS:
            return None
        new_state[row] = p

    for j in range(n):
        if j == row or state[j] == 0:
            continue
        dep = deps[row][j]
        if dep == "x":
            p = state[j] + delta
            if p < MIN_POS or p > MAX_POS:
                return None
            new_state[j] = p
        elif dep == "o":
            p = state[j] - delta
            if p < MIN_POS or p > MAX_POS:
                return None
            new_state[j] = p

    return tuple(new_state)


def build_effects(deps, n):
    """Precompute, per row, the list of ``(index, sign)`` it shifts when moved.

    ``sign`` is +1 for "same direction" (the row itself and 'x' deps) and -1 for
    "opposite" ('o' deps). Doing this once turns each move into a short walk over
    only the affected indices instead of an n-wide rescan of the dependency
    strings on every one of the millions of BFS expansions.
    """
    effects = []
    for r in range(n):
        row_eff = [(r, 1)]
        for j in range(n):
            if j == r:
                continue
            if deps[r][j] == "x":
                row_eff.append((j, 1))
            elif deps[r][j] == "o":
                row_eff.append((j, -1))
        effects.append(row_eff)
    return effects


def _step(state, row, delta, effects):
    """Fast move using precomputed effects. Returns new tuple or ``None``."""
    new_state = list(state)
    for j, sign in effects[row]:
        if state[j] == 0:
            continue
        p = state[j] + sign * delta
        if p < MIN_POS or p > MAX_POS:
            return None
        new_state[j] = p
    return tuple(new_state)


def _reconstruct(parent, end):
    """Walk the parent map back to the start to recover the move list."""
    moves = []
    node = end
    while parent[node] is not None:
        prev, row, direction = parent[node]
        moves.append((row, direction))
        node = prev
    moves.reverse()
    return moves


def solve(state, deps, n, active):
    """Shortest move sequence bringing every ``active`` row to position 4.

    Breadth-first search, so the returned path is guaranteed minimal. Memory is
    O(states): instead of carrying a growing move list with every queued node
    (the old approach -- O(states x path length)), we store one parent pointer
    per discovered state and rebuild the path once at the end.

    Returns the list of ``(row, direction)`` moves, ``[]`` if already solved, or
    ``None`` if unsolvable / the search space cap is exceeded.
    """
    start = tuple(state)
    if all(start[i] == TARGET for i in active):
        return []

    effects = build_effects(deps, n)
    directions = (("left", 1), ("right", -1))

    parent = {start: None}  # state -> (prev_state, row, direction); start -> None
    queue = deque([start])
    max_states = 5_000_000

    while queue:
        cur = queue.popleft()
        for row in range(n):
            for direction, delta in directions:
                nxt = _step(cur, row, delta, effects)
                if nxt is None or nxt in parent:
                    continue
                parent[nxt] = (cur, row, direction)
                if all(nxt[i] == TARGET for i in active):
                    return _reconstruct(parent, nxt)
                queue.append(nxt)
                if len(parent) > max_states:
                    return None
    return None
